nixos/version: Only warn about unset stateVersion if used (lite)

roberth opened the pull request:
###### Description of changes

If a configuration does not use services that depend on the `stateVersion`, it does not need to be set.

This provides an incentive for services not to rely on `stateVersion`, and not to burden users with this.

Currently, it is still required by some modules, so to make a valid `stateVersion`-less configuration you'll have to use `disabledModules`.

This change forms the foundation of #207313, which has breaking changes, while this pr does not.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos/version: Only warn about unset stateVersion if used

If a configuration does not use services that depend on the
stateVersion, it does not need to be set.

This provides an incentive for services not to rely on
stateVersion, and not to burden users with this.

roberth committed:
darwin.builder: Avoid unnecessary dependency on stateVersion

dasJ commented:
What about dropping the default altogether?

roberth commented:
> What about dropping the default altogether?

The default default or the macOS one? The macOS linux builder one seems valuable enough because it adds a bit of explanation.

I'm ok with removing the `default` from `misc/version.nix`, but that's a separate change and a separate discussion.

Gabriella439 commented on nixos/modules/profiles/macos-builder.nix:
```diff
@@ -93,7 +104,23 @@ in
+    Please inspect the trace of the following command to figure out which module
+    has a dependency on stateVersion.
+
+      nix-instantiate --argstr system x86_64-darwin -A darwin.builder --show-trace
+
+    or
+
+      nix-instantiate --argstr system aarch64-darwin -A darwin.builder --show-trace
```
Does the `system` need to be explicitly specified?

Also, I try to use long option names (e.g. `--attr`) when possible (especially if the user will likely just copy-and-paste the command)

```suggestion
      nix-instantiate --attr darwin.builder --show-trace
```

Gabriella439 commented on nixos/modules/profiles/macos-builder.nix:
```diff
@@ -93,7 +104,23 @@ in
+
+      nix-instantiate --argstr system aarch64-darwin -A darwin.builder --show-trace
+
+    If this error occurred while evaluating the static part of the option
```
This paragraph might not be necessary because the macOS builder disables documentation: https://github.com/NixOS/nixpkgs/blob/e52e5e0e6bfb34b2eaba08eda8c55e8b79cb4136/nixos/modules/profiles/macos-builder.nix#L27-L28

Gabriella439 reviewed (commented):
Looks great!  Just a few small suggestions

roberth committed:
nixos/macos-builder: Simplify error message

The `system` argument is usually implied, and not hard to figure out.

Co-authored-by: Gabriella Gonzalez <Gabriel439@gmail.com>

roberth committed:
nixos/macos-builder: Remove unnecessary paragraph

I was considering the broader context of stateVersion, but the
macos builder does not enable documentation, making this paragraph
a distraction.
